### PR TITLE
spi: sam0: use Device Tree for configuration.

### PR DESCRIPTION
**Pull request by nzmichaelh**
Define a Device Tree binding for the SAM0 SPI and use it instead of
Kconfig for enabling / disabling instances

Signed-off-by: Michael Hope <mlhx@google.com>

### CODE REVIEW

**Comment by galak** [comment]
You could merge this with #5731

**Comment by nzmichaelh** [comment]
@galak @tbursztyka this is now ready for review.

**Review by galak — CHANGES_REQUESTED** [review]
We need to rework how we deal w/SPI & UART since they are all just modes of SERCOM.

I think we should have a node per SERCOM, have multiple compatibles in the node.  One for “atmel,sercom”, and one for “atmel,sercom-spi”, “atmel,sercom-i2c” or “atmel,sercom-uart”.  Not sure if we should expand the compats to match the all the modes (CTRLA.MODE) can be set to.

**Comment by galak** [comment]
@nzmichaelh any updates here?

**Comment by nzmichaelh** [comment]
Hey @galak I'm happy to rework this but I'll need some pointers on Device Tree :) The peripheral reset and clock configuration code is basically the same for all SERCOMs.  The peripheral is pretty simple so the actual serial, SPI, etc drivers are dominated by how different the Zephyr API is.

**Comment by nzmichaelh** [comment]
@galak let me know what the changes would look like and I'll start working on them

**Comment by nzmichaelh** [comment]
I've also rebased onto master.

**Comment by nzmichaelh** [comment]
Hi @galak I've rebased this and updated it for the Trinket M0.

re: the many modes of SERCOM, I had a look around the code and couldn't find an example of how to do support multiple modes of one peripheral.  Could you tell me how, else could we land this patch and merge the SERCOM modes in a follow up?

**Comment by galak** [comment]
> Hi @galak I've rebased this and updated it for the Trinket M0.
> 
> re: the many modes of SERCOM, I had a look around the code and couldn't find an example of how to do support multiple modes of one peripheral. Could you tell me how, else could we land this patch and merge the SERCOM modes in a follow up?

I need to play around with this, the straight forward method is to modify the compatible in the board .dts file to set it to uart or spi.  I need to chat with a few people to see if there are examples of how this should be handled.

**Comment by galak** [comment]
Ok, after a little discussion I think the compatible solution is the way to go.  So I think we should have sercom nodes in dts/arm/atmel/samd21.dtsi - w/o any compatible.  In the board.dts we than add the compatible in addition to status = "ok";

So something like:

&sercom0 {
        compatible = "atmel,sam0-uart";
        status = "ok";
        current-speed = <115200>;
};

&sercom4 {
        compatible = "atmel,sam0-spi";
        status = "ok";
};

Make sense?

**Comment by nzmichaelh** [comment]
@galak done.  It turned out pretty clean - the Device Tree generator uses the compatible string as part of the #define, so I didn't need to make any other changes.

File: dts/arm/atmel/samd21.dtsi [hunk 1]
```diff
@@ -61,96 +61,54 @@
 			label = "WATCHDOG_0";
 		};
 
-		sercom0: uart@42000800 {
-			compatible = "atmel,sam0-uart";
+		sercom0: sercom@42000800 {
+			compatible = "atmel,sam0-sercom";
```
Comment (ydamigos) [inline dts/arm/atmel/samd21.dtsi]: I couldn't find this constrain in any yaml file.

**Review by ydamigos — COMMENTED** [review]
DT node inheritance is not clear to me. For example, the interrupt property exists in atmel,sam0-uart.yaml but I cannot find it in atmel,sam0-spi.yaml. Does SAM0 spi driver support interrupt?

File: dts/arm/atmel/samd21.dtsi (hunk 1 above)
Comment (nzmichaelh) [reply dts/arm/atmel/samd21.dtsi]: Fixed.

**Comment by nzmichaelh** [comment]
re: sercom, uart, and spi: the SPI driver doesn't use interrupts as the SAM0 has no FIFO so it's faster and cheaper to use polling.

re: inheritance, it seems the Device Tree generator doesn't support multiple inheritance so, for example, the sam0-uart binding can't inherit from both uart.yaml and atmel,sam0-sercom.yaml.

**Comment by ydamigos** [comment]
>re: sercom, uart, and spi: the SPI driver doesn't use interrupts as the SAM0 has no FIFO so it's faster and cheaper to use polling.
>re: inheritance, it seems the Device Tree generator doesn't support multiple inheritance so, for example, the sam0-uart binding can't inherit from both uart.yaml and atmel,sam0-sercom.yaml.

I propose to remove atmel,sam0-sercom.yaml, `compatible = "atmel,sam0-sercom";` and declare sercomX node in samd21.dtsi: 
```
sercomX: sercom@address {
			reg = <address 0x40>;
			status = "disabled";
			label = "SERCOM0";
};
```
In board's dtsi we add uart as:
```
sercomΧ {
 	compatible = "atmel,sam0-uart";
 	current-speed = <115200>;
 	interrupts = <IRQn 0>;
	status = "ok";
};
```
and SPI as you already did.

In board level, you could use aliases and override label property (e.g. SPI_1, UART_1) to make things more clear.

WDYT?

**Comment by nzmichaelh** [comment]
The SPI HW does support interrupts so leaving them in samd21.dtsi SGTM.  They're just not used as the driver is master only.

**Comment by henrikbrixandersen** [comment]
@ydamigos SGTM. Looking forward to seeing this land in the tree.

**Comment by henrikbrixandersen** [comment]
The SPI loopback config for the Arduino Zero `(tests/drivers/spi/spi_loopback/boards/arduino_zero.conf)` will need to be updated as well (replacing `SPI4` with `SERCOM4`).

**Review by galak — APPROVED** [review]

**Pull request merged**